neuland/shared/TCAViewer.h

```diff
@@ -0,0 +1,251 @@
+            {
+                LOG(error) << "TCAViewer<out>: Cannot reset non-existed TClonesArray!";
+            }
+            mTCA->Clear("C");
+        }
+
+        auto InputInit(FairRootManager* rootMan) -> TClonesArray*
+        {
+            auto* tTCA = dynamic_cast<TClonesArray*>(rootMan->GetObject(mBranchName.data()));
+            if (tTCA == nullptr)
+            {
+                LOG(fatal) << "TCAViewer<in>: No TClonesArray called " << mBranchName
+                           << "could be obtained from the FairRootManager!";
+            }
+            return tTCA;
+        }
+
+        auto OutputInit(FairRootManager* rootMan) -> TClonesArray*
+        {
+            mTCA_out = std::make_unique<TClonesArray>(DataType::Class_Name());
+            rootMan->Register(mBranchName.data(), "", mTCA_out.get(), kTRUE);
+            return mTCA_out.get();
+        }
+
+        [[nodiscard]] auto InputGet() -> decltype(auto)
+        {
+            if (mTCA == nullptr)
+            {
+                LOG(warn) << "TCAViewer<in>: get data from a non-existed TClonesArray!";
+            }
+            auto is_sucess = CheckCapcitySize() ? SetFirstAddr() : true;
+            return is_sucess ? std::span<DataType* const>(mCont, mTCA->GetEntriesFast()) : std::span<DataType* const>();
+        }
+
+        [[nodiscard]] auto OutputGet() -> TCAOutput_SafeGuard<DataType>
+        {
+            if (mSafeGuard)
+            {
+                LOG(fatal) << "TCAViewer<out>: Get() function can only be called once in the scope. ";
+            }
+            else
+            {
+                // make sure this function get called every event
+                LogMCEntryNumber();
+                Clear();
+            }
+            return TCAOutput_SafeGuard<DataType>(this);
+        }
+
+        template <typename... Args>
+        void Emplace_back(Args&&... args)
+        {
+            static_assert(mode == out, "TCAViewer<out>::Emplace_back cannot be called in input mode!");
+            if (mTCA == nullptr)
+            {
+                LOG(error) << "TCAViewer<out>: Cannot emplace back a non-existed TClonesArray!";
+            }
+            new ((*mTCA_out)[mTCA_out->GetEntries()]) DataType(std::forward<Args>(args)...);
+        }
+
+        auto SetFirstAddr() -> bool
+        {
+            auto is_success = false;
+            if (mTCA != nullptr && mTCA->Capacity() != 0)
+            {
+                if (dynamic_cast<DataType*>((*mTCA)[0]) == nullptr)
+                {
+                    LOG(fatal) << "TCAViewer<in>: Cannot convert the element in " << mBranchName << " to the type "
+                               << DataType::Class_Name();
+                }
+                else
+                {
+                    mCont = reinterpret_cast<DataType**>(mTCA->GetObjectRef());
+                    is_success = true;
+                }
+            }
+            return is_success;
+        }
+
+        void LogMCEntryNumber()
```

Comment:
`Log...` is not what this does. This is about ensuring the function was called, right? So what about `EnsureFunctionCalled` or similar?
Then, you could pull it together with the safeguard, and pull out the `if safeguard` plus the`EnsureFunctionCalled` from `OutputGet` into a `EnsureFunctionCalledExactlyOnce`.

I don't understand why it is necessary, by the way. :-)

Reply:
```C++
[[nodiscard]] auto OutputGet() -> TCAOutput_SafeGuard<DataType>
{
    EnsureFunctionCalledExactlyOnce();
    Clear();
    return TCAOutput_SafeGuard<DataType>(this);
}
```

Reply:
Yes. The reason is when user put this Get() (or OutputGet) function call inside an if-statement or there is a throw before the function call, Get() function won't be called for every event. And in some events, if function is not called, old data from last call can't be cleared and as the result they would be written down to files for the second time in the end of event loop. This is bad. (I also think clearing data cache in TClonesArray should be handled by FairRootManger, not by ourself.)

So I need to make sure the OutputGet() function must be called at least once to clear the old data. And I also don't want user to create multiple unnecessary output handlers for a single output data. The function should be called exactly once per event loop. Yeah, `EnsureCalledOnce` is a better name.